### PR TITLE
Add tests for PARSE-KEY

### PR DESCRIPTION
**Pull request by PuercoPop**
Also setup the necessary machinery to run the tests through ASDF's
test-op:

  (asdf:test-system "stumpwm")

While trying to rewrite parse-key I kept running into edge-cases. I've added some happy path examples as well as some edge cases to test parse-key to get a test suite going.

Something I'm unsure of, although it makes sense conceptually for the test-op to be on StumpWM, as it currently stands every time we change a file on StumpWM the test-op will recompile the system before running the tests, which are not immediate.

Let me know what you think

Ref. #453

### CODE REVIEW

File: stumpwm-tests.asd [hunk 1]
```diff
@@ -0,0 +1,8 @@
+(defsystem "stumpwm-tests"
```
Comment (sjl) [inline stumpwm-tests.asd]: If you don't want to add a separate `.asd` file you can name the system with a slash `stumpwm/tests` and put this in the main `.asd` and ASDF will be smart enough to look there when you try to load it, e.g. https://github.com/sjl/bobbin/blob/master/bobbin.asd

More info in the last paragraph of `find-system` in the ASDF manual: https://common-lisp.net/project/asdf/asdf.html#index-find_002dsystem

**Comment by joaotavora** [comment]
> I'm fairly certain that it is used by SISCOG, where Luismbo and Joao works (or used to idk?).

Yes, I can confirm that.

> Fiasco does a good job hiding them if I don't want to use them, but I'd rather remove them from the code.

Open an issue and propose an alternative (though @luismbo would probably want to weigh in on that too).

> And although I originally like the concept of a test being a function, If I can't reuse the same deftest with different parameters there is little to gain from them being functions.

`trace` support and find-definition are two useful things (I don't know if you get that with any other test suites).  When an error or test failure happens, you also get a readable backtrace where your test appears as its own frame.   And you get autodoc too (but if you don't use arguments then I guess It's not super useful).

File: stumpwm-tests.asd (hunk 1 above)
Comment (PuercoPop) [reply stumpwm-tests.asd]: I'm aware, but I prefer the one file per project. I remember there was a bug with the logic handling the / in ASDF recently (that has already been fixed) where it treated the slash as a sub-directory. I rather make it 'easier' for ASDF.

**Comment by sdilts** [comment]
Would it be a good idea to run the tests with travis? If we have tests, we might as well run them when we are checking if the project builds or not.

File: stumpwm.asd [hunk 2]
```diff
@@ -59,4 +59,7 @@
                (:file "remap-keys")
                ;; keep this last so it always gets recompiled if
                ;; anything changes
-               (:file "version")))
+               (:file "version"))
+  :in-order-to ((test-op (load-op "stumpwm-tests")))
+  :perform (test-op (o c)
+                    (uiop/package:symbol-call "FIASCO" "RUN-TESTS" 'stumpwm-tests)))
```
Comment (jorams) [inline stumpwm.asd]: Instead of adding this `:perform` to the `stumpwm` system, I think it should be in the `stumpwm-tests` system. Then the option above it would be:

```lisp
:in-order-to ((test-op (test-op "stumpwm-tests")))
```

Comment (PuercoPop) [reply stumpwm.asd]: Thanks, updated the PR

**Comment by PuercoPop** [comment]
I had forgotten about Parachute. I've never used it but my experience Shinmera's software has always been well documented and thought out. I wound't mind using it either.

@dbjergaard Let me know what's your final call on the test framework.

<hr />

@jorams afaiu, the reason for Prove's (plan ...) is that it is required by the TAP protocol. It is used as a heuristic to detect when your test suite has hanged. Useful when running on the CI.

**Comment by dbjergaard** [comment]
Fiasco is fine. I will merge this PR soon. Can someone make sure we don’t need to update the build instructions to reflect the new dependency on fiasco?

    David

**Comment by PuercoPop** [comment]
@dbjergaard We don't need to update the update the new build instructions because Fiasco is a dependency of a new system, "stumpwm-tests", not of Stumpwm itself. We should update the travis instructions as @sdilts mentioned though.
